File: src/test/java/com/cattailsw/nanidroid/runtime/GhostPresentationReducerTest.java
package com.cattailsw.nanidroid.runtime;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

/** Locks the pure presentation boundary that the legacy view adapter consumes. */
public final class GhostPresentationReducerTest {
    @Test
    public void requiredMigrationInvariant_snapshotPreservesSpeakerTextSurfaceAnimationAndBalloonPolicy() {
        GhostPresentationState state = GhostPresentationReducer.snapshot(
                "Sakura text", "120", "3", "-1",
                "Kero text", "11", null, "0");

        assertEquals("Sakura text", state.getSakura().getText());
        assertEquals("120", state.getSakura().getSurfaceId());
        assertEquals("3", state.getSakura().getAnimationId());
        assertTrue(state.getSakura().getBalloonVisible());

        assertEquals("Kero text", state.getKero().getText());
        assertEquals("11", state.getKero().getSurfaceId());
        assertEquals(null, state.getKero().getAnimationId());
        assertTrue(state.getKero().getBalloonVisible());
    }

    @Test
    public void requiredMigrationInvariant_emptyTextAndDisabledBalloonRemainHidden() {
        GhostPresentationState state = GhostPresentationReducer.snapshot(
                "", "0", null, "-1",
                "", "10", null, "-1");

        assertFalse(state.getSakura().getBalloonVisible());
        assertFalse(state.getKero().getBalloonVisible());
    }
}

File: src/test/java/com/cattailsw/nanidroid/runtime/GhostPresentationReducerTest.kt
package com.cattailsw.nanidroid.runtime

import org.junit.Assert
import org.junit.Test

/** Locks the pure presentation boundary that the legacy view adapter consumes.  */
class GhostPresentationReducerTest {
    @Test
    fun requiredMigrationInvariant_snapshotPreservesSpeakerTextSurfaceAnimationAndBalloonPolicy() {
        val state = GhostPresentationReducer.snapshot(
            "Sakura text", "120", "3", "-1",
            "Kero text", "11", null, "0"
        )

        Assert.assertEquals("Sakura text", state.sakura.text)
        Assert.assertEquals("120", state.sakura.surfaceId)
        Assert.assertEquals("3", state.sakura.animationId)
        Assert.assertTrue(state.sakura.balloonVisible)

        Assert.assertEquals("Kero text", state.kero.text)
        Assert.assertEquals("11", state.kero.surfaceId)
        Assert.assertEquals(null, state.kero.animationId)
        Assert.assertTrue(state.kero.balloonVisible)
    }

    @Test
    fun requiredMigrationInvariant_emptyTextAndDisabledBalloonRemainHidden() {
        val state = GhostPresentationReducer.snapshot(
            "", "0", null, "-1",
            "", "10", null, "-1"
        )

        Assert.assertFalse(state.sakura.balloonVisible)
        Assert.assertFalse(state.kero.balloonVisible)
    }
}